File: 14-TypeScript/Grey Multi-Tenant/packages/grey-elixir/lib/grey_sdk/error/grey_error.ex
defmodule GreySdk.Error do
  @moduledoc """
  Normalized error structure for all Grey SDK operations.

  All errors from the Grey SDK are normalized into this structure,
  providing consistent error handling across all operations.

  ## Fields

  - `code` - Error code identifying the type of error
  - `message` - Human-readable error message
  - `details` - Optional map with additional error details

  ## Common Error Codes

  - `"UNAUTHORIZED"` - Authentication required or failed
  - `"FORBIDDEN"` - Permission denied
  - `"NOT_FOUND"` - Resource not found
  - `"VALIDATION_ERROR"` - Invalid input
  - `"NETWORK_ERROR"` - Network connectivity issue
  - `"TIMEOUT"` - Request timed out
  - `"SERVER_ERROR"` - Internal server error
  - `"UNKNOWN"` - Unknown error

  ## Examples

      case GreySdk.Auth.login(client, email, password) do
        {:ok, auth} ->
          {:ok, auth}

        {:error, %GreySdk.Error{code: "UNAUTHORIZED"} = error} ->
          Logger.warning("Auth failed: \#{error.message}")
          {:error, :unauthorized}

        {:error, %GreySdk.Error{code: code, message: message}} ->
          Logger.error("Error [\#{code}]: \#{message}")
          {:error, :unknown}
      end
  """

  @type t :: %__MODULE__{
          code: String.t(),
          message: String.t(),
          details: map() | nil
        }

  @enforce_keys [:code, :message]
  defstruct [:code, :message, :details]

  # Common error codes
  @unauthorized "UNAUTHORIZED"
  @forbidden "FORBIDDEN"
  @not_found "NOT_FOUND"
  @validation_error "VALIDATION_ERROR"
  @network_error "NETWORK_ERROR"
  @timeout "TIMEOUT"
  @server_error "SERVER_ERROR"
  @unknown "UNKNOWN"

  @doc "Returns the UNAUTHORIZED error code."
  def unauthorized_code, do: @unauthorized

  @doc "Returns the FORBIDDEN error code."
  def forbidden_code, do: @forbidden

  @doc "Returns the NOT_FOUND error code."
  def not_found_code, do: @not_found

  @doc "Returns the VALIDATION_ERROR error code."
  def validation_error_code, do: @validation_error

  @doc "Returns the NETWORK_ERROR error code."
  def network_error_code, do: @network_error

  @doc "Returns the TIMEOUT error code."
  def timeout_code, do: @timeout

  @doc "Returns the SERVER_ERROR error code."
  def server_error_code, do: @server_error

  @doc "Returns the UNKNOWN error code."
  def unknown_code, do: @unknown

  @doc """
  Creates an unauthorized error.

  ## Examples

      iex> GreySdk.Error.unauthorized()
      %GreySdk.Error{code: "UNAUTHORIZED", message: "Not authenticated"}

      iex> GreySdk.Error.unauthorized("Invalid credentials")
      %GreySdk.Error{code: "UNAUTHORIZED", message: "Invalid credentials"}
  """
  @spec unauthorized(String.t()) :: t()
  def unauthorized(message \\ "Not authenticated") do
    %__MODULE__{code: @unauthorized, message: message}
  end

  @doc """
  Creates a validation error.

  ## Examples

      iex> GreySdk.Error.validation("Email is required")
      %GreySdk.Error{code: "VALIDATION_ERROR", message: "Email is required"}
  """
  @spec validation(String.t()) :: t()
  def validation(message) do
    %__MODULE__{code: @validation_error, message: message}
  end

  @doc """
  Creates a not found error.
  """
  @spec not_found(String.t()) :: t()
  def not_found(message \\ "Resource not found") do
    %__MODULE__{code: @not_found, message: message}
  end

  @doc """
  Creates a network error.
  """
  @spec network(String.t()) :: t()
  def network(message \\ "Network error") do
    %__MODULE__{code: @network_error, message: message}
  end

  @doc """
  Creates a timeout error.
  """
  @spec timeout(String.t()) :: t()
  def timeout(message \\ "Request timed out") do
    %__MODULE__{code: @timeout, message: message}
  end

  @doc """
  Creates an unknown error.
  """
  @spec unknown(String.t()) :: t()
  def unknown(message \\ "Unknown error") do
    %__MODULE__{code: @unknown, message: message}
  end

  @doc """
  Creates a GreyError from a gRPC error.

  Maps gRPC status codes to Grey error codes.
  """
  @spec from_grpc_error(GRPC.RPCError.t()) :: t()
  def from_grpc_error(%GRPC.RPCError{status: status, message: message}) do
    code = grpc_status_to_code(status)
    %__MODULE__{code: code, message: message || "gRPC error", details: %{grpc_status: status}}
  end

  def from_grpc_error(error) when is_exception(error) do
    %__MODULE__{code: @unknown, message: Exception.message(error)}
  end

  def from_grpc_error(error) do
    %__MODULE__{code: @unknown, message: inspect(error)}
  end

  @doc """
  Creates a GreyError from any term.
  """
  @spec from_any(any()) :: t()
  def from_any(%__MODULE__{} = error), do: error
  def from_any(%GRPC.RPCError{} = error), do: from_grpc_error(error)
  def from_any(error) when is_exception(error), do: %__MODULE__{code: @unknown, message: Exception.message(error)}
  def from_any(error) when is_binary(error), do: %__MODULE__{code: @unknown, message: error}
  def from_any(error), do: %__MODULE__{code: @unknown, message: inspect(error)}

  # Maps gRPC status codes to Grey error codes
  defp grpc_status_to_code(status) do
    case status do
      :unauthenticated -> @unauthorized
      :permission_denied -> @forbidden
      :not_found -> @not_found
      :invalid_argument -> @validation_error
      :failed_precondition -> @validation_error
      :unavailable -> @network_error
      :deadline_exceeded -> @timeout
      :internal -> @server_error
      :unknown -> @unknown
      _ -> @unknown
    end
  end
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-erlang/src/grey_error.erl
%%%-------------------------------------------------------------------
%%% @doc Grey SDK error handling module.
%%%
%%% Provides normalized error structures and utilities for all Grey SDK
%%% operations.
%%% @end
%%%-------------------------------------------------------------------
-module(grey_error).

-include("grey_sdk.hrl").

%% API exports
-export([
    new/2,
    new/3,
    unauthorized/0,
    unauthorized/1,
    validation/1,
    not_found/0,
    not_found/1,
    network/0,
    network/1,
    timeout/0,
    timeout/1,
    server_error/0,
    server_error/1,
    unknown/0,
    unknown/1,
    from_grpc_error/1,
    from_any/1,
    code/1,
    message/1,
    details/1,
    to_map/1
]).

%% Error code constants
-define(UNAUTHORIZED, <<"UNAUTHORIZED">>).
-define(FORBIDDEN, <<"FORBIDDEN">>).
-define(NOT_FOUND, <<"NOT_FOUND">>).
-define(VALIDATION_ERROR, <<"VALIDATION_ERROR">>).
-define(NETWORK_ERROR, <<"NETWORK_ERROR">>).
-define(TIMEOUT, <<"TIMEOUT">>).
-define(SERVER_ERROR, <<"SERVER_ERROR">>).
-define(UNKNOWN, <<"UNKNOWN">>).

%%====================================================================
%% API functions
%%====================================================================

%% @doc Creates a new error with code and message.
-spec new(binary(), binary()) -> #grey_error{}.
new(Code, Message) ->
    #grey_error{code = Code, message = Message, details = undefined}.

%% @doc Creates a new error with code, message, and details.
-spec new(binary(), binary(), map()) -> #grey_error{}.
new(Code, Message, Details) ->
    #grey_error{code = Code, message = Message, details = Details}.

%% @doc Creates an unauthorized error.
-spec unauthorized() -> #grey_error{}.
unauthorized() ->
    unauthorized(<<"Not authenticated">>).

-spec unauthorized(binary()) -> #grey_error{}.
unauthorized(Message) ->
    new(?UNAUTHORIZED, Message).

%% @doc Creates a validation error.
-spec validation(binary()) -> #grey_error{}.
validation(Message) ->
    new(?VALIDATION_ERROR, Message).

%% @doc Creates a not found error.
-spec not_found() -> #grey_error{}.
not_found() ->
    not_found(<<"Resource not found">>).

-spec not_found(binary()) -> #grey_error{}.
not_found(Message) ->
    new(?NOT_FOUND, Message).

%% @doc Creates a network error.
-spec network() -> #grey_error{}.
network() ->
    network(<<"Network error">>).

-spec network(binary()) -> #grey_error{}.
network(Message) ->
    new(?NETWORK_ERROR, Message).

%% @doc Creates a timeout error.
-spec timeout() -> #grey_error{}.
timeout() ->
    timeout(<<"Request timed out">>).

-spec timeout(binary()) -> #grey_error{}.
timeout(Message) ->
    new(?TIMEOUT, Message).

%% @doc Creates a server error.
-spec server_error() -> #grey_error{}.
server_error() ->
    server_error(<<"Internal server error">>).

-spec server_error(binary()) -> #grey_error{}.
server_error(Message) ->
    new(?SERVER_ERROR, Message).

%% @doc Creates an unknown error.
-spec unknown() -> #grey_error{}.
unknown() ->
    unknown(<<"Unknown error">>).

-spec unknown(binary()) -> #grey_error{}.
unknown(Message) ->
    new(?UNKNOWN, Message).

%% @doc Creates a GreyError from a gRPC error.
-spec from_grpc_error(term()) -> #grey_error{}.
from_grpc_error({grpc_error, Status, Message}) ->
    Code = grpc_status_to_code(Status),
    new(Code, to_binary(Message), #{grpc_status => Status});
from_grpc_error({error, {grpc_error, Status, Message}}) ->
    from_grpc_error({grpc_error, Status, Message});
from_grpc_error(Other) ->
    from_any(Other).

%% @doc Creates a GreyError from any term.
-spec from_any(term()) -> #grey_error{}.
from_any(#grey_error{} = Error) ->
    Error;
from_any({grpc_error, _, _} = Error) ->
    from_grpc_error(Error);
from_any(Error) when is_binary(Error) ->
    unknown(Error);
from_any(Error) when is_list(Error) ->
    unknown(list_to_binary(Error));
from_any(Error) when is_atom(Error) ->
    unknown(atom_to_binary(Error, utf8));
from_any(Error) ->
    unknown(list_to_binary(io_lib:format("~p", [Error]))).

%% @doc Gets the error code.
-spec code(#grey_error{}) -> binary().
code(#grey_error{code = Code}) ->
    Code.

%% @doc Gets the error message.
-spec message(#grey_error{}) -> binary().
message(#grey_error{message = Message}) ->
    Message.

%% @doc Gets the error details.
-spec details(#grey_error{}) -> map() | undefined.
details(#grey_error{details = Details}) ->
    Details.

%% @doc Converts error to a map.
-spec to_map(#grey_error{}) -> map().
to_map(#grey_error{code = Code, message = Message, details = Details}) ->
    Map = #{code => Code, message => Message},
    case Details of
        undefined -> Map;
        _ -> Map#{details => Details}
    end.

%%====================================================================
%% Internal functions
%%====================================================================

%% Maps gRPC status codes to Grey error codes.
grpc_status_to_code(unauthenticated) -> ?UNAUTHORIZED;
grpc_status_to_code(permission_denied) -> ?FORBIDDEN;
grpc_status_to_code(not_found) -> ?NOT_FOUND;
grpc_status_to_code(invalid_argument) -> ?VALIDATION_ERROR;
grpc_status_to_code(failed_precondition) -> ?VALIDATION_ERROR;
grpc_status_to_code(unavailable) -> ?NETWORK_ERROR;
grpc_status_to_code(deadline_exceeded) -> ?TIMEOUT;
grpc_status_to_code(internal) -> ?SERVER_ERROR;
grpc_status_to_code(unknown) -> ?UNKNOWN;
grpc_status_to_code(_) -> ?UNKNOWN.

to_binary(V) when is_binary(V) -> V;
to_binary(V) when is_list(V) -> list_to_binary(V);
to_binary(V) when is_atom(V) -> atom_to_binary(V, utf8);
to_binary(V) -> list_to_binary(io_lib:format("~p", [V])).
